crates/ghost_actor/src/ghost_tracker.rs

```diff
@@ -0,0 +1,393 @@
+            }
+        }
+
+        for request_id in expired {
+            match self.pending.remove(&request_id) {
+                None => (),
+                Some(entry) => {
+                    (entry.cb)(user_data, Err("timeout".into()))?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct GhostTrackerBookmarkOptions {
+    pub timeout_ms: Option<u64>,
+}
+
+impl Default for GhostTrackerBookmarkOptions {
+    fn default() -> Self {
+        Self { timeout_ms: None }
+    }
+}
+
+impl GhostTrackerBookmarkOptions {
+    pub fn set_timeout_ms(mut self, timeout_ms: u64) -> Self {
+        self.timeout_ms = Some(timeout_ms);
+        self
+    }
+}
+
+/// GhostTracker registers callbacks associated with request_ids
+/// that can be triggered later when a response comes back indicating that id
+pub struct GhostTracker<'lt, X: 'lt + Send + Sync, T: 'lt + Send + Sync> {
+    // just for ref count
+    _inner: Arc<GhostMutex<GhostTrackerInner<'lt, X, T>>>,
+    send_inner: crossbeam_channel::Sender<GhostTrackerToInner<'lt, X, T>>,
+}
+
+impl<'lt, X: 'lt + Send + Sync, T: 'lt + Send + Sync> GhostTracker<'lt, X, T> {
+    pub fn new(mut sys_ref: GhostSystemRef<'lt>, weak_user_data: Weak<GhostMutex<X>>) -> Self {
+        let (send_inner, recv_inner) = crossbeam_channel::unbounded();
+
+        let inner = Arc::new(GhostMutex::new(GhostTrackerInner::new(
+            sys_ref.clone(),
+            weak_user_data,
+            recv_inner,
+        )));
+        let weak_inner = Arc::downgrade(&inner);
+
+        sys_ref
+            .enqueue_processor(
+                0,
+                Box::new(move || match weak_inner.upgrade() {
+                    Some(strong_inner) => {
+                        let mut strong_inner = strong_inner.lock();
+                        match strong_inner.weak_user_data.upgrade() {
+                            Some(strong_user_data) => {
+                                let mut strong_user_data = strong_user_data.lock();
+
+                                strong_inner
+                                    .process(&mut *strong_user_data)
+                                    .expect("tracker process error");
+
+                                Ok(GhostProcessInstructions::default().set_should_continue(true))
+                            }
+                            None => {
+                                // we don't have any user_data, next time?
+                                Ok(GhostProcessInstructions::default().set_should_continue(true))
+                            }
+                        }
+                    }
+                    None => Ok(GhostProcessInstructions::default()),
+                }),
+            )
+            .expect("can enqueue processor");
+
+        Self {
+            _inner: inner,
+            send_inner,
+        }
+    }
+
+    /// register a periodic task
+    pub fn periodic_task(
```

Comment:
Suggest renaming to `add_periodic_task()`

Reply:
ended up pulling out the periodic thing for now... updated https://github.com/holochain/lib3h/issues/288 with a hint about this name